README.md
[![Build Status](http://venus:8080/view/Integration%20Jobs/job/I40-rbdl-dubnium-devel_dubnium/badge/icon)](http://venus:8080/view/Integration%20Jobs/job/I40-rbdl-dubnium-devel_dubnium/)

RBDL - Rigid Body Dynamics Library
Copyright (c) 2011-2016 Martin Felis <martin.felis@iwr.uni-heidelberg.de>

Introduction
============

RBDL is a highly efficient C++ library that contains some essential rigid
body dynamics algorithms such as the Articulated Body Algorithm (ABA) for
forward dynamics, Recursive Newton-Euler Algorithm (RNEA) for inverse
dynamics and the Composite Rigid Body Algorithm (CRBA) for the efficient
computation of the joint space inertia matrix. It further contains code for
Jacobians, forward and inverse kinematics, and handling of external
constraints such as contacts and collisions.

The code is developed by Martin Felis <martin.felis@iwr.uni-heidelberg.de>
at the research group [Optimization in Robotics and Biomechanics
(ORB)](http://orb.iwr.uni-heidelberg.de) of the [Interdisciplinary Center
for Scientific Computing (IWR)](http://www.iwr.uni-heidelberg.de) at
[Heidelberg University](http://www.uni-heidelberg.de). The code tightly
follows the notation used in Roy Featherstone''s book "Rigid Body Dynamics
Algorithm".

Recent Changes
==============
  * 28 April 2016: Nev version 2.5.0:
    * Added an experimental Cython based Python wrapper of RBDL. The API is
      very close to the C++ API. For a brief glimpse of the API see the file
      python/test_wrapper.py.
    * Matthew Millard added CustomJoints which allow to create different joint
      types completely by user code. They are implemented as proxy joints for
      which their behaviour is specified using virtual functions.
    * Added CalcMInvTimesTau() that evaluates multiplication of the inverse of
      the joint space inertia matrix with a vector in O(n) time.
    * Added JointTypeFloatingBase which uses TX,TY,TZ and a spherical joint for
      the floating base joint.
    * Loading of floating base URDF models must now be specified as a third
      parameter to URDFReadFromFile() and URDFReadFromString()
    * Added the URDF code from Bullet3 which gets used when ROS is not found.
      Otherwise use the URDF libraries found via Catkin.
    * Added CalcPointVelocity6D, CalcPointAcceleration6D, and CalcPointJacobian6D
      that compute both linear and angular quantities
    * Removed Model::SetFloatingBase (body). Use a 6-DoF joint or
      JointTypeFloatingBase instead.
    * Fixed building issues when building DLL with MSVC++.
  * 20 April 2016: New version 2.4.1:
    * This is a bugfix release that maintains binary compatibility and only fixes
    erroneous behaviour.
    * critical: fixed termination criterion for InverseKinematics. The termination
      criterion would be evaluated too early and thus report convergence too
      early. This was reported independently by Kevin Stein, Yun Fei, and Davide
      Corradi. Thanks for the reports!
    * critical: fixed CompositeRigidBodyAlgorithm when using spherical joints
      (thanks to Sébastien Barthélémy for reporting!)
  * 23 February 2015: New version 2.4.0:
    * Added sparse range-space method ForwardDynamicsContactsRangeSpaceSparse()
      and ComputeContactImpulsesRangeSpaceSparse() 
    * Added null-space method ForwardDynamicsContactsNullSpace()
      and ComputeContactImpulsesNullSpace() 
    * Renamed ForwardDynamicsContactsLagrangian() to
      ForwardDynamicsContactsDirect() and
      ComputeContactImpulsesLagrangian() to ComputeContactImpulsesDirect()
    * Renamed ForwardDynamicsContacts() to ForwardDynamicsContactsKokkevis()
    * Removed/Fixed CalcAngularMomentum(). The function produced wrong values. The
      functionality has been integrated into CalcCenterOfMass().
    * CalcPointJacobian() does not clear the argument of the result anymore.
      Caller has to ensure that the matrix was set to zero before using this
      function.
    * Added optional workspace parameters for ForwardDynamicsLagrangian() to
      optionally reduce memory allocations
    * Added JointTypeTranslationXYZ, JointTypeEulerXYZ, and JointTypeEulerYXZ
      which are equivalent to the emulated multidof joints but faster.
    * Added optional parameter to CalcCenterOfMass to compute angular momentum.
    * Added CalcBodySpatialJacobian()
    * Added CalcContactJacobian()
    * Added NonlinearEffects()
    * Added solving of linear systems using standard Householder QR
    * LuaModel: Added LuaModelReadFromLuaState()
    * URDFReader: Fixed various issues and using faster joints for floating
      base models
    * Various performance improvements

For a complete history see doc/api_changes.txt.

Documentation
=============

The documentation is contained in the code and can be extracted with the
tool [doxygen](http://www.doxygen.org).

To create the documentation simply run

    doxygen Doxyfile

which will generate the documentation in the subdirectory ./doc/html. The
main page will then be located in ./doc/html/index.html.

An online version of the generated documentation can be found at
[http://rbdl.bitbucket.org](http://rbdl.bitbucket.org).

Getting RBDL
============

The latest stable code can be obtained from

    https://bitbucket.org/rbdl/rbdl/get/default.zip

The official mercurial repository can be cloned from

    https://bitbucket.org/rbdl/rbdl

(See [http://mercurial.selenic.com/](http://mercurial.selenic.com) for
mercurial clients.)

Building and Installation
=========================

The RBDL is built using CMake
([http://www.cmake.org](http://www.cmake.org)). To compile the library in
a separate directory in Release mode use:

    mkdir build
    cd build/
    cmake -D CMAKE_BUILD_TYPE=Release ../ 
    make

For optimal performance it is highly recommended to install the Eigen3
linear algebra library from
[http://eigen.tuxfamily.org](http://eigen.tuxfamily.org). RBDL also
comes with a simple, albeit much slower math library (SimpleMath) that can
be used by enabling `RBDL_USE_SIMPLE_MATH`, i.e.:

    cmake -D RBDL_USE_SIMPLE_MATH=TRUE ../

Licensing
=========

The library is published under the very permissive zlib free software
license which should allow you to use the software wherever you need. 

This is the full license text (zlib license):

    RBDL - Rigid Body Dynamics Library
    Copyright (c) 2011-2016 Martin Felis <martin.felis@iwr.uni-heidelberg.de>
    
    This software is provided 'as-is', without any express or implied
    warranty. In no event will the authors be held liable for any damages
    arising from the use of this software.
    
    Permission is granted to anyone to use this software for any purpose,
    including commercial applications, and to alter it and redistribute it
    freely, subject to the following restrictions:
    
       1. The origin of this software must not be misrepresented; you must not
       claim that you wrote the original software. If you use this software
       in a product, an acknowledgment in the product documentation would be
       appreciated but is not required.
    
       2. Altered source versions must be plainly marked as such, and must not
       be misrepresented as being the original software.
    
       3. This notice may not be removed or altered from any source
       distribution.

Acknowledgements
================

Work on this library was funded by the [Heidelberg Graduate School of
Mathematical and Computational Methods for the Sciences
(HGS)](http://hgs.iwr.uni-heidelberg.de/hgs.mathcomp/), and the European
FP7 projects [ECHORD](http://echord.eu) (grant number 231143) and
[Koroibot](http://koroibot.eu) (grant number 611909).


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package rbdl
^^^^^^^^^^^^^^^^^^^^^^^^^^

0.4.1 (2023-10-24)
------------------
* Merge branch 'gripper_additional_mass' into 'erbium-devel'
  Add API to update body inertial parameters at runtime -
  See merge request control/rbdl!21
* Model: move new methods to update bodies inertia to public
* Fix differences between main rbdl branch and PAL's fork
* updated intertial api function names to comply with the rest of rbdl function names
* Add unit test to change inertia of fixed bodies in model
* Methods to update body inertia now correctly work for fixed bodies
* Add unit test to check the update of body inertial parameters
* Add methods in Model to update body inertial parameters correctly
* Add unit test for Body::Separate method
* Factorize duplicated code beetwen Body::Join and Body::Separate in another method
* Add a Body::Separate method to remove the effect of a body inertial parameters that was previously joined
* Add method Model::updateInertiaMatrixForBody
* Merge branch 'centroial_inertia_matrix' into 'erbium-devel'
  added the methods to compute the spatial and centroidal inertia matrix
  See merge request control/rbdl!19
* added more tests for computation of spatial inertia and momentum
* added a new argument to compute the angular momentum and also updated tests
* added tests for computation of Inertia of the body
* use mBodies instead of the mBodyNameMap
* added the methods to compute the spatial and centroidal inertia matrix
* Contributors: Adria Roig, Pierre Fernbach, Sai Kishor Kothakota, ju6ge

0.4.0 (2022-04-08)
------------------
* Merge branch 'dynamic_linearized_mpc' into 'erbium-devel'
  Add global inertia matrix from COM method
  See merge request control/rbdl!17
* Merge branch 'update_ik_api_from_upstream' into 'erbium-devel'
  Update new InverseKinematics with different constraints API
  See merge request control/rbdl!18
* Update new InverseKinematics with different constraints API
* Add global inertia matrix from COM method
* Contributors: Adria Roig, Sai Kishor Kothakota, saikishor

0.3.0 (2021-08-29)
------------------
* Merge branch 'kangaroo_wbc' into 'erbium-devel'
  Compute ZMP kinetimatically and partiall mass methods
  See merge request control/rbdl!16
* Compute ZMP kinetimatically and partiall mass methods
* Contributors: Adria Roig, victor

0.2.12 (2019-11-06)
-------------------
* Merge branch 'gcc-warning' into 'erbium-devel'
  Gcc warning
  See merge request control/rbdl!15
* disable false positives errors on tests
* Disable shadowed variable
* Disable gcc uninitialized
* Contributors: Victor Lopez

0.2.11 (2019-09-05)
-------------------
* Merge branch 'melodic-devel' into 'erbium-devel'
  Melodic devel
  See merge request control/rbdl!14
* Fix build errors
* Use urdf compatibility typedefs for kinetic/melodic compatibility
* Fix float comparison
* Add missing SYSTEM to include_directories
* migrated to melodic
* Contributors: Hilario Tome, Victor Lopez

0.2.10 (2019-07-24)
-------------------
* Merge branch 'inertia_id' into 'erbium-devel'
  Inertia
  See merge request control/rbdl!13
* cleanup of the relative stuff
* adding relative quantities
* added relative computations
* Contributors: Hilario Tome

0.2.9 (2019-03-29)
------------------
* Merge branch 'clang-fixes' into 'erbium-devel'
  Wrong usage of template keyword
  See merge request control/rbdl!12
* Wrong usage of template keyword
* Contributors: Victor Lopez

0.2.8 (2019-03-18)
------------------
* Merge branch 'com_template' into 'erbium-devel'
  Com template
  See merge request control/rbdl!11
* fixed critical bug in forward dynamics with multidof joint
* templetized some functions
* Merge branch 'faster-getbodyid' into 'erbium-devel'
  Twice as fast getbodyid
  See merge request control/rbdl!8
* Twice as fast getbodyid
* Contributors: Hilario Tome, Victor Lopez

0.2.7 (2018-11-15)
------------------
* Merge branch 'clang-fixes' into 'erbium-devel'
  Fix wrong usage of dependent template
  See merge request control/rbdl!9
* Fix wrong usage of dependent template
* Contributors: Victor Lopez

0.2.6 (2018-07-06)
------------------
* Merge branch 'as_test_fix' into 'erbium-devel'
  Fix uninitialized values in urdfreader.
  See merge request control/rbdl!7
* Fix uninitialized values in urdfreader.
* Contributors: Hilario Tome, alexandersherikov

0.2.5 (2018-06-08)
------------------
* Merge branch 'continuous_joint' into 'erbium-devel'
  changed how continous joint limit is parsed
  See merge request control/rbdl!5
* Fix: velocity bounds for continuous joints
* changed how continous joint limit is parsed
* Contributors: Hilario Tome, alexandersherikov

0.2.4 (2018-04-17)
------------------
* Merge branch 'pose_offset_test' into 'erbium-devel'
  Add pose offset API
  See merge request control/rbdl!3
* Clang format
* Fix rebase issues
* Fix compilling issues
* Add pose offset API
* Contributors: Adria Roig, Hilario Tome

0.2.3 (2018-04-06)
------------------
* Merge branch 'link_reference_to' into 'erbium-devel'
  added get transform with respect to other link
  See merge request control/rbdl!4
* added get transform with respect to other link
* Contributors: Hilario Tome

0.2.2 (2018-02-14)
------------------
* Merge branch 'as-urdfreader-linkskip' into 'erbium-devel'
  urdfreader: added functions to read URDF without given links
  See merge request control/rbdl!1
* Merge branch 'erbium-devel' into as-urdfreader-linkskip
* Dummy commit to trigger recursive testing
* Use quaternions instead of RPY angles to set joint transformations.
  Previous version seem to be introducing noise due to conversions back
  and forth (quaternion -> RPY -> rotation matrix).
* Model: added const getModelData().
* URDF reader: allow explicit specification of the kinamatic root link.
* URDF reader: minor changes in the logic
* Minor bugfix in URDF reader.
* Omit links in URDFModel: deleted corresponding functions from URDF reader
* urdfreader: code deduplication & cleanup

ggtest/MathTests.cpp
#include <gtest/gtest.h>
#include <eigen_checks/gtest.h>
#include "rbdl/Logging.h"
#include "rbdl/rbdl_math.h"
#include "rbdl/rbdl_mathutils.h"
#include <iostream>

const double TEST_PREC = 1.0e-14;

using namespace std;
using namespace RigidBodyDynamics::Math;

class MathFixture : public ::testing::Test{
protected:
  virtual void SetUp(){}
  virtual void TearDown(){}
};

TEST (MathTests, GaussElimPivot) {
  ClearLogOutput();

  MatrixNd A;
  A.resize(3,3);
  VectorNd b(3);
  VectorNd x(3);

  A(0,0) = 0; A(0,1) = 2; A(0,2) = 1;
  A(1,0) = 1; A(1,1) = 1; A(1,2) = 5;
  A(2,0) = 0; A(2,1) = 0; A(2,2) = 1;

  b[0] = 1;
  b[1] = 2;
  b[2] = 3;

  VectorNd test_result (3);

  test_result[0] = -12;
  test_result[1] = -1;
  test_result[2] = 3;

  LinSolveGaussElimPivot (A, b, x);

  EXPECT_TRUE(EIGEN_MATRIX_NEAR (test_result, x, TEST_PREC));

  A(0,0) = 0; A(0,1) = -2; A(0,2) = 1;
  A(1,0) = 1; A(1,1) =  1; A(1,2) = 5;
  A(2,0) = 0; A(2,1) =  0; A(2,2) = 1;

  LinSolveGaussElimPivot (A, b, x);
  test_result[0] = -14;
  test_result[1] = 1;
  test_result[2] = 3;

  EXPECT_TRUE(EIGEN_MATRIX_NEAR (test_result, x,  TEST_PREC));
}

TEST (MathTests, Dynamic_1D_initialize_value) {
  VectorNd myvector_10 = VectorNd::Constant ((size_t) 10, 12.);
  VectorNd test_values(10);

  for (unsigned int i = 0; i < 10; i++)
    test_values[i] = 12.;

  EXPECT_TRUE(EIGEN_MATRIX_EQUAL_DOUBLE(test_values, myvector_10));
}

TEST (MathTests,Dynamic_2D_initialize_value) {
  MatrixNd mymatrix_10x10 = MatrixNd::Constant (10, 10, 12.);
  MatrixNd test_values(10, 10);

  for (unsigned int i = 0; i < 10; i++)
    for (unsigned int j = 0; j < 10; j++)
      test_values(i,j) = 12.;

  EXPECT_TRUE(EIGEN_MATRIX_EQUAL_DOUBLE(test_values, mymatrix_10x10));
}

TEST (MathTests, SpatialMatrix_Multiplication) {
  SpatialMatrixd X_1 (
      1.,  2.,  3.,  4.,  5.,  6.,
      11., 12., 13., 14., 15., 16.,
      21., 22., 23., 24., 25., 26.,
      31., 32., 33., 34., 35., 36.,
      41., 42., 43., 44., 45., 46.,
      51., 52., 53., 54., 55., 56.
      );

  SpatialMatrixd X_2 (X_1);

  X_2 *= 2;

  SpatialMatrixd correct_result (
      1442,    1484,    1526,    1568,    1610,    1652,
      4562,    4724,    4886,    5048,    5210,    5372,
      7682,    7964,    8246,    8528,    8810,    9092,
      10802,   11204,   11606,   12008,   12410,   12812,
      13922,   14444,   14966,   15488,   16010,   16532,
      17042,   17684,   18326,   18968,   19610,   20252
      );

  SpatialMatrixd test_result = X_1 * X_2;

  EXPECT_TRUE(EIGEN_MATRIX_NEAR (correct_result, test_result, TEST_PREC));

  // check the *= operator:
  test_result = X_1;
  test_result *= X_2;

  EXPECT_TRUE(EIGEN_MATRIX_NEAR (correct_result, test_result,  TEST_PREC));
}

int main( int argc, char** argv)
{
  testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}
